src/Validation/Rule/FieldOnCorrectTypeRule.php

```diff
@@ -0,0 +1,137 @@
+<?php
+
+namespace Digia\GraphQL\Validation\Rule;
+
+use Digia\GraphQL\Error\ValidationException;
+use Digia\GraphQL\Language\AST\Node\FieldNode;
+use Digia\GraphQL\Language\AST\Node\InterfacesTrait;
+use Digia\GraphQL\Language\AST\Node\NodeInterface;
+use Digia\GraphQL\Type\Definition\AbstractTypeInterface;
+use Digia\GraphQL\Type\Definition\FieldsTrait;
+use Digia\GraphQL\Type\Definition\InterfaceType;
+use Digia\GraphQL\Type\Definition\NameTrait;
+use Digia\GraphQL\Type\Definition\ObjectType;
+use Digia\GraphQL\Type\Definition\OutputTypeInterface;
+use Digia\GraphQL\Type\Definition\TypeInterface;
+use Digia\GraphQL\Type\SchemaInterface;
+use function Digia\GraphQL\Util\suggestionList;
+use function Digia\GraphQL\Validation\undefinedFieldMessage;
+
+/**
+ * Fields on correct type
+ *
+ * A GraphQL document is only valid if all fields selected are defined by the
+ * parent type, or are an allowed meta field such as __typename.
+ */
+class FieldOnCorrectTypeRule extends AbstractRule
+{
+    /**
+     * @inheritdoc
+     */
+    public function enterNode(NodeInterface $node): ?NodeInterface
+    {
+        if ($node instanceof FieldNode) {
```

Comment:
I would return early here to save one indent level

Reply:
Well yes, @hungneox said the same. However, if you look at the rest of the rules I think it's easier to understand if we leave it like this.